File: module2/class1/exemplo12.cpp
/*
================================================================================
ATIVIDADE PRÁTICA 12 - RESOURCE LEAKS E RAII (C++)
================================================================================

OBJETIVO:
- Demonstrar vazamentos de recursos do sistema (handles, sockets)
- Usar system resource monitors para identificar leaks
- Otimizar usando RAII (Resource Acquisition Is Initialization)
- Medir impacto de resource leaks no sistema

PROBLEMA:
- Não fechar handles/recursos causa system resource leaks
- malloc/new sem free/delete causa memory leaks
- Resource monitoring mostrará crescimento de handles

SOLUÇÃO:
- Usar RAII pattern com smart pointers
- Automatic cleanup em destructors

================================================================================
*/

#include <iostream>
#include <chrono>
#include <vector>
#include <fstream>
#include <memory>
using namespace std;

void demonstrateResourceLeaks() {
    cout << "Starting resource leak demonstration..." << endl;
    cout << "Monitor system resources - should see growing resource usage" << endl;
    
    const int ITERATIONS = 500;
    vector<FILE*> leakedFiles;
    vector<int*> leakedMemory;
    
    auto start = chrono::high_resolution_clock::now();
    
    for (int i = 0; i < ITERATIONS; i++) {
        // PERFORMANCE ISSUE: Opening files without closing them
        FILE* file = fopen("temp.txt", "w");
        if (file) {
            fprintf(file, "Test data %d", i);
            fflush(file);
            leakedFiles.push_back(file);
            // Missing fclose(file) - file handle leak!
        }
        
        // PERFORMANCE ISSUE: Allocating memory without freeing
        int* buffer = new int[1000];
        for (int j = 0; j < 1000; j++) {
            buffer[j] = i * j;
        }
        leakedMemory.push_back(buffer);
        // Missing delete[] buffer - memory leak!
        
        // PERFORMANCE ISSUE: Creating ofstream without proper cleanup
        ofstream* stream = new ofstream("temp2.txt");
        *stream << "Data: " << i << endl;
        // Missing delete stream and stream->close() - resource leak!
        
        if (i % 50 == 0) {
            cout << "Created " << i << "/" << ITERATIONS << " leaked resources..." << endl;
        }
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end - start);
    
    cout << "Resource leak test completed in: " << duration.count() << " ms" << endl;
    cout << "File handles leaked: " << leakedFiles.size() << endl;
    cout << "Memory blocks leaked: " << leakedMemory.size() << endl;
    cout << "Stream objects leaked: " << ITERATIONS << endl;
}

int main() {
    cout << "Starting resource management demonstration..." << endl;
    cout << "Task: Creating system resources without proper cleanup" << endl;
    cout << "Monitor system resources and memory usage" << endl << endl;
    
    demonstrateResourceLeaks();
    
    cout << endl << "=== PROFILING ANALYSIS ===" << endl;
    cout << "Check resource monitors for:" << endl;
    cout << "- Growing file handle count" << endl;
    cout << "- Memory usage increase" << endl;
    cout << "- System resource consumption" << endl;
    
    return 0;
}

/*
================================================================================
VERSÃO CORRIGIDA (DESCOMENTE PARA USAR RAII PATTERN)
================================================================================

#include <iostream>
#include <chrono>
#include <vector>
#include <fstream>
#include <memory>
using namespace std;

// CORREÇÃO: RAII wrapper for FILE*
class FileWrapper {
    FILE* file;
public:
    explicit FileWrapper(const char* filename, const char* mode) : file(fopen(filename, mode)) {}
    
    ~FileWrapper() {
        if (file) {
            fclose(file); // Automatic cleanup
        }
    }
    
    FILE* get() const { return file; }
    bool valid() const { return file != nullptr; }
};

void demonstrateRAII() {
    cout << "Starting RAII demonstration..." << endl;
    cout << "Monitor system resources - should remain stable" << endl;
    
    const int ITERATIONS = 500;
    
    auto start = chrono::high_resolution_clock::now();
    
    for (int i = 0; i < ITERATIONS; i++) {
        // CORREÇÃO: RAII ensures automatic file closure
        {
            FileWrapper file("temp.txt", "w");
            if (file.valid()) {
                fprintf(file.get(), "Test data %d", i);
                fflush(file.get());
            }
        } // Automatic fclose() in destructor
        
        // CORREÇÃO: Smart pointer ensures automatic memory cleanup
        {
            auto buffer = make_unique<int[]>(1000);
            for (int j = 0; j < 1000; j++) {
                buffer[j] = i * j;
            }
        } // Automatic delete[] in unique_ptr destructor
        
        // CORREÇÃO: Stack-allocated ofstream with automatic cleanup
        {
            ofstream stream("temp2.txt");
            stream << "Data: " << i << endl;
        } // Automatic close() and destructor cleanup
        
        if (i % 50 == 0) {
            cout << "Processed " << i << "/" << ITERATIONS << " resources with RAII..." << endl;
        }
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end - start);
    
    cout << "RAII demonstration completed in: " << duration.count() << " ms" << endl;
    cout << "No resource leaks - all resources automatically cleaned up" << endl;
}

int main() {
    cout << "Starting optimized RAII demonstration..." << endl;
    cout << "Task: Creating system resources with automatic cleanup" << endl;
    cout << "Monitor system resources for stable usage" << endl << endl;
    
    demonstrateRAII();
    
    cout << endl << "=== OPTIMIZATION RESULTS ===" << endl;
    cout << "Improvements:" << endl;
    cout << "- No resource leaks due to RAII pattern" << endl;
    cout << "- Automatic cleanup in destructors" << endl;
    cout << "- Exception-safe resource management" << endl;
    cout << "- Better system resource utilization" << endl;
    
    return 0;
}

================================================================================
*/

File: module2/class1/exemplo12.cs
/*
================================================================================
ATIVIDADE PRÁTICA 12 - RESOURCE LEAKS E DISPOSE PATTERNS (C#)
================================================================================

OBJETIVO:
- Demonstrar vazamentos de recursos não-gerenciados
- Usar Memory profiler para identificar handles/recursos não liberados
- Otimizar usando using statements e Dispose patterns
- Medir impacto de resource leaks no sistema

PROBLEMA:
- Não fazer Dispose() de IDisposable causa resource leaks
- Handles de sistema não são liberados automaticamente
- Resource monitoring mostrará crescimento de handles

SOLUÇÃO:
- Usar using statements para cleanup automático
- Implementar Dispose pattern corretamente

================================================================================
*/

using System;
using System.IO;
using System.Diagnostics;

class Program {
    static void DemonstrateResourceLeaks() {
        Console.WriteLine("Starting resource leak demonstration...");
        Console.WriteLine("Monitor system resources - should see growing handle count");
        
        const int ITERATIONS = 1000;
        
        var sw = Stopwatch.StartNew();
        
        for (int i = 0; i < ITERATIONS; i++) {
            // PERFORMANCE ISSUE: Creating FileStream without proper disposal
            var fs = new FileStream("temp.txt", FileMode.Create, FileAccess.Write);
            fs.Write(new byte[] { 1, 2, 3, 4, 5 });
            fs.Flush();
            // Missing fs.Dispose() - resource leak!
            
            // PERFORMANCE ISSUE: Creating Process without disposal
            var process = new Process();
            process.StartInfo = new ProcessStartInfo("cmd.exe", "/c echo test")
            {
                UseShellExecute = false,
                CreateNoWindow = true
            };
            // Missing process.Dispose() - handle leak!
            
            if (i % 100 == 0) {
                Console.WriteLine($"Created {i}/{ITERATIONS} undisposed resources...");
                Console.WriteLine($"Current process handles: {Process.GetCurrentProcess().HandleCount}");
            }
        }
        
        sw.Stop();
        
        Console.WriteLine($"Resource leak test completed in: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Final handle count: {Process.GetCurrentProcess().HandleCount}");
        Console.WriteLine($"Resources created without disposal: {ITERATIONS * 2}");
        
        GC.Collect();
        GC.WaitForPendingFinalizers();
        GC.Collect();
        
        Console.WriteLine($"Handle count after GC: {Process.GetCurrentProcess().HandleCount}");
    }
    
    static void Main() {
        Console.WriteLine("Starting resource management demonstration...");
        Console.WriteLine("Task: Creating system resources without proper cleanup");
        Console.WriteLine("Monitor system resources and handle counts");
        Console.WriteLine();
        
        DemonstrateResourceLeaks();
        
        Console.WriteLine();
        Console.WriteLine("=== PROFILING ANALYSIS ===");
        Console.WriteLine("Check resource monitors for:");
        Console.WriteLine("- Growing handle count");
        Console.WriteLine("- File handles not released");
        Console.WriteLine("- System resource consumption");
    }
}

/*
================================================================================
VERSÃO CORRIGIDA (DESCOMENTE PARA USAR A VERSÃO COM PROPER DISPOSAL)
================================================================================

using System;
using System.IO;
using System.Diagnostics;

class Program {
    static void DemonstrateProperResourceManagement() {
        Console.WriteLine("Starting proper resource management demonstration...");
        Console.WriteLine("Monitor system resources - handle count should remain stable");
        
        const int ITERATIONS = 1000;
        int initialHandles = Process.GetCurrentProcess().HandleCount;
        
        var sw = Stopwatch.StartNew();
        
        for (int i = 0; i < ITERATIONS; i++) {
            // CORREÇÃO: Using statement ensures proper disposal
            using (var fs = new FileStream("temp.txt", FileMode.Create, FileAccess.Write)) {
                fs.Write(new byte[] { 1, 2, 3, 4, 5 });
                fs.Flush();
            } // Automatic fs.Dispose() called here
            
            // CORREÇÃO: Using statement for Process disposal
            using (var process = new Process()) {
                process.StartInfo = new ProcessStartInfo("cmd.exe", "/c echo test")
                {
                    UseShellExecute = false,
                    CreateNoWindow = true
                };
                // Process.Dispose() automatically called
            }
            
            if (i % 100 == 0) {
                Console.WriteLine($"Created {i}/{ITERATIONS} properly disposed resources...");
                Console.WriteLine($"Current process handles: {Process.GetCurrentProcess().HandleCount}");
            }
        }
        
        sw.Stop();
        
        Console.WriteLine($"Proper resource management completed in: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Initial handles: {initialHandles}");
        Console.WriteLine($"Final handle count: {Process.GetCurrentProcess().HandleCount}");
        Console.WriteLine($"Handle growth: {Process.GetCurrentProcess().HandleCount - initialHandles}");
        
        // Clean up test file
        if (File.Exists("temp.txt")) {
            File.Delete("temp.txt");
        }
    }
    
    static void Main() {
        Console.WriteLine("Starting optimized resource management demonstration...");
        Console.WriteLine("Task: Creating system resources with proper cleanup");
        Console.WriteLine("Monitor system resources for stable handle usage");
        Console.WriteLine();
        
        DemonstrateProperResourceManagement();
        
        Console.WriteLine();
        Console.WriteLine("=== OPTIMIZATION RESULTS ===");
        Console.WriteLine("Improvements:");
        Console.WriteLine("- Stable handle count due to proper disposal");
        Console.WriteLine("- No resource leaks");
        Console.WriteLine("- Deterministic cleanup with using statements");
        Console.WriteLine("- Better system resource utilization");
    }
}

================================================================================
*/
